### PR TITLE
feat: add support for the safe navigation (aka Elvis) operator

### LINKED ISSUE #791: Elvis operator in expressions: `?.`
Currently `a.b.c` will thrown an error in expression if `a` is `null`. To suppress null dereferencing we should support `a?.b.c` which would short circuit further evaluation if `a` is null. In this way we could have explicit notation which would show which values could be `null`.

Syntax to fallow dart proposal: https://github.com/gbracha/nullAwareOperators/blob/master/README.md

### PR DESCRIPTION
**Pull request by vicb**
fixes #791

### CODE REVIEW

**Comment by vicb** [comment]
@kegluneq depending on which of your Dart PR or this one will get merged first, one of us will have to update the Dart JIT code to add safe methods.

**Comment by kegluneq** [comment]
Thanks for the heads up -- we'll have to update the Js JIT code also, which is modified a decent bit in my PR.

File: modules/angular2/src/change_detection/parser/lexer.ts [hunk 1]
```diff
@@ -219,15 +219,16 @@ class _Scanner {
       case $DQ:
         return this.scanString();
       case $HASH:
-        return this.scanOperator(start, StringWrapper.fromCharCode(peek));
+        return this.scanOperator(start, '#');
```
Comment (IgorMinar) [inline modules/angular2/src/change_detection/parser/lexer.ts]: why don't you just let it fall through and let $CARET handle it?

Comment (vicb) [inline modules/angular2/src/change_detection/parser/lexer.ts]: good catch, I'll update

File: modules/angular2/test/change_detection/parser/lexer_spec.js [hunk 2]
```diff
@@ -162,7 +162,7 @@ export function main() {
         expectOperatorToken(tokens[8], 22, '!==');
       });
 
-      it('should tokenize statements', function() {
+      it('should tokenize statements', () => {
```
Comment (IgorMinar) [inline modules/angular2/test/change_detection/parser/lexer_spec.js]: these changes are unrelated to the intent of this PR. next time please create a separate commit for style fixes

Comment (IgorMinar) [inline modules/angular2/test/change_detection/parser/lexer_spec.js]: the reason being that you are bloating up the PR and in the future when someone is looking at blame log for a file they'll see a lot of unrelated noise.
